File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.cpp
class Solution {
public:
    int pivotindex(vector<int>& nums,int n){
        int start=0,end=nums.size()-1;
        int mid=start+(end-start)/2;
        while(start<end){
            if(nums[mid]>=nums[0]){
                start=start+1;
            }
            else{
                end=mid;
            }
            mid=start+(end-start)/2;
        }
        return start;
    }
    int binarysearch(vector<int>& nums,int low,int high,int k){
        int start=low,end=high;
        int mid=start+(end-start)/2;
        //cout<<"low:"<<low<<"high:"<<end<<endl;
        while(start<=end){
            if(nums[mid]==k){
                return mid;
            }
            else if(nums[mid]<k){
                start=mid+1;
            }
            else
            {
                end=mid - 1;
            }
            mid=start+(end-start)/2;
           // cout<<"mid:"<<mid<<endl;
        }
        return -1;
    }
    int search(vector<int>& nums, int target) {
        int output=-1;
        //int mid=low+(end-low)/2;
        int i=pivotindex(nums,nums.size());
        cout<<i<<endl;
        if(target>=nums[i] && target<=nums[nums.size()-1]){
            output=binarysearch(nums,i,nums.size()-1,target);
        }
        else{
            output=binarysearch(nums,0,i-1,target);
        }
        //cout<<"output:"<<output<<endl;
        return output;
    }
};

File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.java
class Solution {
    public int pivotElement(int[] nums){
        int low=0;
        int high=nums.length-1;
        int mid=low+(high-low)/2;
        while(low<=high){
            if(low==high)return mid;
            if(mid>0 && nums[mid-1]>nums[mid])return mid-1;
            else if(nums[mid]>nums[mid+1])return mid;
            else if(nums[mid]>=nums[low])low=mid+1;
            else high=mid;
            mid=low+(high-low)/2;
        }
        return high;
    }
    public int bs(int[] nums,int low,int high, int target){
        int mid=low+(high-low)/2;
        while(low<=high){
            if(nums[mid]==target)return mid;
            else if (nums[mid]<target)low=mid+1;
            else high=mid-1;
            mid=low+(high-low)/2;
        }
        return -1;
    }
    public int search(int[] nums, int target) {
        int pivot=pivotElement(nums);
       
        if(nums[pivot]>target && nums[0]>target){
            return bs(nums,pivot+1,nums.length-1,target);
        }
        return  bs(nums,0,pivot,target);
    }
}